File: frontend/lib/enums.ts
// lib/enums.ts
// Frontend mirror of app/models/enums.py — values must stay byte-identical
// to the backend enums. Single source of truth for status/role strings on
// the frontend; import these everywhere instead of writing string literals.

export const UserRole = {
  client: "client",
  broker: "broker",
  admin: "admin",
} as const;
export type UserRole = (typeof UserRole)[keyof typeof UserRole];

export const OTPPurpose = {
  login: "login",
  signup: "signup",
  broker_verification: "broker_verification",
} as const;
export type OTPPurpose = (typeof OTPPurpose)[keyof typeof OTPPurpose];

export const PropertyStatus = {
  draft: "draft",
  pending: "pending",
  active: "active",
  sold: "sold",
  rented: "rented",
  expired: "expired",
  rejected: "rejected",
} as const;
export type PropertyStatus = (typeof PropertyStatus)[keyof typeof PropertyStatus];

export const ListingType = {
  sale: "sale",
  rent: "rent",
  pg: "pg",
} as const;
export type ListingType = (typeof ListingType)[keyof typeof ListingType];

// Human-readable labels for each ListingType, shared by PropertyHeader
// (Details page) and the homepage Listings section's card badge. Callers
// that also need per-badge color styling keep their own wrapper around
// this label.
export const LISTING_TYPE_LABELS: Record<ListingType, string> = {
  [ListingType.sale]: "For Sale",
  [ListingType.rent]: "For Rent",
  [ListingType.pg]: "PG / Co-living",
};

export const PropertyType = {
  apartment: "apartment",
  villa: "villa",
  independent_house: "independent_house",
  plot: "plot",
  land: "land",
  office: "office",
  shop: "shop",
  pg_colive: "pg_colive",
  // Rent-only, per the Post Property wizard's "Commercial & Spaces" dropdown group.
  commercial_building: "commercial_building",
  built_to_suit: "built_to_suit",
} as const;
export type PropertyType = (typeof PropertyType)[keyof typeof PropertyType];

// Human-readable labels for each PropertyType, shared by the Listings
// filter sidebar and the Comparison table so both render identical
// category names.
export const PROPERTY_TYPE_LABELS: Record<PropertyType, string> = {
  [PropertyType.apartment]: "Apartment",
  [PropertyType.villa]: "Villa",
  [PropertyType.independent_house]: "Independent House",
  [PropertyType.plot]: "Plot",
  [PropertyType.land]: "Land",
  [PropertyType.office]: "Office",
  [PropertyType.shop]: "Shop",
  [PropertyType.pg_colive]: "PG / Co-living",
  [PropertyType.commercial_building]: "Commercial Building",
  [PropertyType.built_to_suit]: "Built to Suit",
};

export const PriceFlexibility = {
  fixed: "fixed",
  negotiable: "negotiable",
  highly_flexible: "highly_flexible",
} as const;
export type PriceFlexibility = (typeof PriceFlexibility)[keyof typeof PriceFlexibility];

export const PRICE_FLEXIBILITY_LABELS: Record<PriceFlexibility, string> = {
  [PriceFlexibility.fixed]: "Fixed Price",
  [PriceFlexibility.negotiable]: "Negotiable",
  [PriceFlexibility.highly_flexible]: "Highly Flexible",
};

export const PaymentStructure = {
  full_payment: "full_payment",
  emi_installments: "emi_installments",
  construction_linked: "construction_linked",
} as const;
export type PaymentStructure = (typeof PaymentStructure)[keyof typeof PaymentStructure];

export const PAYMENT_STRUCTURE_LABELS: Record<PaymentStructure, string> = {
  [PaymentStructure.full_payment]: "Full Payment",
  [PaymentStructure.emi_installments]: "EMI / Installments",
  [PaymentStructure.construction_linked]: "Construction Linked",
};

export const Furnishing = {
  unfurnished: "unfurnished",
  semi_furnished: "semi_furnished",
  fully_furnished: "fully_furnished",
} as const;
export type Furnishing = (typeof Furnishing)[keyof typeof Furnishing];

export const MediaType = {
  image: "image",
  video: "video",
} as const;
export type MediaType = (typeof MediaType)[keyof typeof MediaType];

export const LeadStatus = {
  new: "new",
  contacted: "contacted",
  site_visit: "site_visit",
  negotiation: "negotiation",
  closed_won: "closed_won",
  closed_lost: "closed_lost",
} as const;
export type LeadStatus = (typeof LeadStatus)[keyof typeof LeadStatus];

export const VerificationStatus = {
  unverified: "unverified",
  pending: "pending",
  verified: "verified",
  rejected: "rejected",
} as const;
export type VerificationStatus = (typeof VerificationStatus)[keyof typeof VerificationStatus];

export const BoostTier = {
  basic: "basic",
  featured: "featured",
  premium: "premium",
} as const;
export type BoostTier = (typeof BoostTier)[keyof typeof BoostTier];

export const OrderStatus = {
  created: "created",
  paid: "paid",
  failed: "failed",
  refunded: "refunded",
} as const;
export type OrderStatus = (typeof OrderStatus)[keyof typeof OrderStatus];

export const BrokerDocumentType = {
  rera_certificate: "rera_certificate",
  government_id: "government_id",
} as const;
export type BrokerDocumentType = (typeof BrokerDocumentType)[keyof typeof BrokerDocumentType];

File: backend/app/models/enums.py
"""
app/models/enums.py

Single source of truth for every domain enum used across the models.
Each enum is mirrored as a native Postgres enum type in the Alembic
migrations. Import these enums everywhere a status, role, or category
value is needed instead of using raw strings.
"""

from enum import Enum


class UserRole(str, Enum):
    """Account type: determines portal access and permissions."""

    client = "client"
    broker = "broker"
    admin = "admin"


class PropertyStatus(str, Enum):
    """Lifecycle stage of a property listing."""

    draft = "draft"
    pending = "pending"
    active = "active"
    sold = "sold"
    rented = "rented"
    expired = "expired"
    rejected = "rejected"


class ListingType(str, Enum):
    """Transaction type a property is listed under."""

    sale = "sale"
    rent = "rent"
    pg = "pg"


class PropertyType(str, Enum):
    """Physical category of a property."""

    apartment = "apartment"
    villa = "villa"
    independent_house = "independent_house"
    plot = "plot"
    land = "land"
    office = "office"
    shop = "shop"
    pg_colive = "pg_colive"


class Furnishing(str, Enum):
    """Furnishing level of a property."""

    unfurnished = "unfurnished"
    semi_furnished = "semi_furnished"
    fully_furnished = "fully_furnished"


class VerificationStatus(str, Enum):
    """Verification state of a broker's identity/credentials."""

    unverified = "unverified"
    pending = "pending"
    verified = "verified"
    rejected = "rejected"


class LeadStatus(str, Enum):
    """Stage of a lead within a broker's sales pipeline."""

    new = "new"
    contacted = "contacted"
    site_visit = "site_visit"
    negotiation = "negotiation"
    closed_won = "closed_won"
    closed_lost = "closed_lost"


class NotificationType(str, Enum):
    """Category of an in-app notification, used to route/render it."""

    property_alert = "property_alert"
    price_drop = "price_drop"
    lead_received = "lead_received"
    message = "message"
    listing_status = "listing_status"
    system = "system"


class AlertFrequency(str, Enum):
    """How often a saved property alert should be delivered."""

    instant = "instant"
    daily = "daily"
    weekly = "weekly"


class BoostTier(str, Enum):
    """Paid listing-boost tier."""

    basic = "basic"
    featured = "featured"
    premium = "premium"


class OrderStatus(str, Enum):
    """Payment status of a boost order."""

    created = "created"
    paid = "paid"
    failed = "failed"
    refunded = "refunded"


class MediaType(str, Enum):
    """Type of a property media asset."""

    image = "image"
    video = "video"


class OTPPurpose(str, Enum):
    """Reason an OTP code was issued."""

    login = "login"
    signup = "signup"
    broker_verification = "broker_verification"


class BrokerDocumentType(str, Enum):
    """
    Kind of a broker verification document. Not a Postgres enum type —
    it only ever appears as the "type" field inside broker_profiles.
    verification_documents (JSONB), never as a table column, so there's
    nothing to migrate.
    """

    rera_certificate = "rera_certificate"
    government_id = "government_id"
